### DISCUSSION TITLE
How to set the window url in rust at runtime?

### DISCUSSION THREAD

**Discussion by jourdain** [tauri-apps/tauri#5377] [Q&A] [answered]
Hi All,

Even if I'm struggling to wrap my head around rust, I've been enjoying tauri. 
Thanks for making such great product!

My current simple use case is that I start a server (via sidecar) on some random port and I would like the WebView to connect to it once I know where that server is running. I've started from the splash screen example to get 2 windows while being able to switch between them.
But now I'm struggling on how to set the window URL from Rust, therefore I'm wondering if someone could help me out.

My current code is below and seems to just be missing just that `set_url` call...

```rust
#![cfg_attr(
  all(not(debug_assertions), target_os = "windows"),
  windows_subsystem = "windows"
)]

use tauri::api::process::{Command, CommandEvent};
use tauri::Manager;

fn main() {
  tauri::Builder::default()
    .setup(|app| {
      let splashscreen_window = app.get_window("splashscreen").unwrap();
      let main_window = app.get_window("main").unwrap();

      let (mut rx, _) = Command::new_sidecar("server")
        .expect("failed to create sidecar")
        .spawn()
        .expect("Failed to spawn server");

      tauri::async_runtime::spawn(async move {
        while let Some(event) = rx.recv().await {
          if let CommandEvent::Stdout(line) = event {
            if line.contains("Network: ") {
              let tokens: Vec<&str> = line.split("Network: ").collect();
              println!("{}", tokens[1]);

              splashscreen_window.close().unwrap();
              main_window.show().unwrap(); // .set_url(tokens[1])
            }
          }
        }
      });
      Ok(())
    })
    .run(tauri::generate_context!())
    .expect("error while running application");
}
```

**Comment by FabianLars** [comment, marked as answer]
There indeed is no direct api for that (yet), so instead you need to use `eval` to execute js code. Something like this:
```rs
main_window.eval(&format!("window.location.replace('http://localhost:{}')", "1234"));
```

**Comment by jourdain** [comment]
Awesome, that's working thanks!

**Comment by jessetabak** [comment]
Seems that after changing the location you lose access to the Tauri Command invoker from window.\_\_TAURI\_\_.tauri
Any idea how to get those back?

(Use case: I want to load my web application from my own remote web server, but still be able to invoke Tauri Commands.)

**Reply by FabianLars** [reply]
If you don't have to change the location but can spawn the window with the remote location you can set the url as your distDir and it should have access to tauri's apis.

if you have (server-side) redirects you need to wait for https://github.com/tauri-apps/tauri/pull/5918

**Comment by jourdain** [comment]
Tauri Commands are only available when content is served by tauri itself.

**Comment by jessetabak** [comment]
Thats too bad, was hoping to ditch Electron this time around. Thanks for the quick reply though, much appreciated.

**Comment by danielfaust** [comment]
But it should be possible to load a page into an iframe which can then communicate with the containing page which in turn should have access to Tauri commands and act as a bridge. Does this work?

**Reply by jourdain** [reply]
Maybe, but I can't tell for sure since I did not try it.

**Comment by yujonglee** [comment]
https://docs.rs/tauri/2.3.1/tauri/webview/struct.Webview.html#method.navigate
